### PR TITLE
Log Flow activity

### PR DESCRIPTION
**Pull request by blizzz**
* Extends ILogger to log custom data
* Logs flow activiation by event hooks (i.e. complex operators that have own listeners like Autotagging and accesscontrol are below radar)
* Logs rule fetching and evaluation  (any Operation is caught here)
* Logs to $datadir/flow.log by default (can be disabled via cli, or a different location specified)
* No flow app has to do anything itself 
* Rotates according to general rotation logic

Example content:

```
{"reqId":"X5XP4vROpcGCoaE6Ria2","level":1,"time":"2020-01-15T09:32:29+00:00","remoteAddr":"127.0.0.1","user":"master","app":"workflowengine","method":"GET","url":"/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping","message":{"operation":{"class":"OCA\\Talk\\Flow\\Operation","name":"Write to conversation"},"entity":{"class":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","name":"Web Request"},"eventName":"FlowHttpRequest::incomingRequest","description":"Flow activated by event"},"userAgent":"curl/7.68.0","version":"18.0.0.8"}
{"reqId":"X5XP4vROpcGCoaE6Ria2","level":1,"time":"2020-01-15T09:32:29+00:00","remoteAddr":"127.0.0.1","user":"master","app":"workflowengine","method":"GET","url":"/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping","message":{"scopes":[{"scope":"admin"},{"scope":"user","uid":"master"}],"entity":{"class":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","name":"Web Request"},"operation":{"class":"OCA\\Talk\\Flow\\Operation","name":"Write to conversation"},"description":"Flow activation: rules were requested"},"userAgent":"curl/7.68.0","version":"18.0.0.8"}
{"reqId":"X5XP4vROpcGCoaE6Ria2","level":1,"time":"2020-01-15T09:32:29+00:00","remoteAddr":"127.0.0.1","user":"master","app":"workflowengine","method":"GET","url":"/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping","message":{"entity":{"class":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","name":"Web Request"},"operation":{"class":"OCA\\Talk\\Flow\\Operation","name":"Write to conversation"},"configuration":{"id":60,"class":"OCA\\Talk\\Flow\\Operation","name":"","checks":"[81]","operation":"{\"m\":3,\"t\":\"wfwqkvsc\"}","events":"[\"FlowHttpRequest::incomingRequest\"]","entity":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","scope_type":1,"scope_actor_id":"master"},"description":"Flow rule qualified to run"},"userAgent":"curl/7.68.0","version":"18.0.0.8"}
{"reqId":"X5XP4vROpcGCoaE6Ria2","level":1,"time":"2020-01-15T09:32:29+00:00","remoteAddr":"127.0.0.1","user":"master","app":"workflowengine","method":"GET","url":"/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping","message":{"entity":{"class":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","name":"Web Request"},"operation":{"class":"OCA\\Talk\\Flow\\Operation","name":"Write to conversation"},"configuration":[{"id":60,"class":"OCA\\Talk\\Flow\\Operation","name":"","checks":"[81]","operation":"{\"m\":3,\"t\":\"wfwqkvsc\"}","events":"[\"FlowHttpRequest::incomingRequest\"]","entity":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","scope_type":1,"scope_actor_id":"master"}],"description":"All qualified flow configurations are going to run"},"userAgent":"curl/7.68.0","version":"18.0.0.8"}
{"reqId":"X5XP4vROpcGCoaE6Ria2","level":1,"time":"2020-01-15T09:32:29+00:00","remoteAddr":"127.0.0.1","user":"master","app":"workflowengine","method":"GET","url":"/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping","message":{"operation":{"class":"OCA\\Talk\\Flow\\Operation","name":"Write to conversation"},"entity":{"class":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","name":"Web Request"},"eventName":"FlowHttpRequest::incomingRequest","description":"Flow handling for event done"},"userAgent":"curl/7.68.0","version":"18.0.0.8"}
{"reqId":"X5XP4vROpcGCoaE6Ria2","level":1,"time":"2020-01-15T09:32:29+00:00","remoteAddr":"127.0.0.1","user":"master","app":"workflowengine","method":"GET","url":"/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping","message":{"operation":{"class":"OCA\\WorkflowScript\\Operation","name":"External scripts"},"entity":{"class":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","name":"Web Request"},"eventName":"FlowHttpRequest::incomingRequest","description":"Flow activated by event"},"userAgent":"curl/7.68.0","version":"18.0.0.8"}
{"reqId":"X5XP4vROpcGCoaE6Ria2","level":1,"time":"2020-01-15T09:32:29+00:00","remoteAddr":"127.0.0.1","user":"master","app":"workflowengine","method":"GET","url":"/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping","message":{"operation":{"class":"OCA\\WorkflowScript\\Operation","name":"External scripts"},"entity":{"class":"OCA\\FlowHttpRequests\\Flow\\RequestEntity","name":"Web Request"},"eventName":"FlowHttpRequest::incomingRequest","description":"Flow handling for event done"},"userAgent":"curl/7.68.0","version":"18.0.0.8"}
```
<details>
<summary>:lipstick: Prettified</summary>

`cat data/flow.log  | jq`

```json
{
  "reqId": "X5XP4vROpcGCoaE6Ria2",
  "level": 1,
  "time": "2020-01-15T09:32:29+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "master",
  "app": "workflowengine",
  "method": "GET",
  "url": "/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping",
  "message": {
    "operation": {
      "class": "OCA\\Talk\\Flow\\Operation",
      "name": "Write to conversation"
    },
    "entity": {
      "class": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
      "name": "Web Request"
    },
    "eventName": "FlowHttpRequest::incomingRequest",
    "description": "Flow activated by event"
  },
  "userAgent": "curl/7.68.0",
  "version": "18.0.0.8"
}
{
  "reqId": "X5XP4vROpcGCoaE6Ria2",
  "level": 1,
  "time": "2020-01-15T09:32:29+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "master",
  "app": "workflowengine",
  "method": "GET",
  "url": "/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping",
  "message": {
    "scopes": [
      {
        "scope": "admin"
      },
      {
        "scope": "user",
        "uid": "master"
      }
    ],
    "entity": {
      "class": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
      "name": "Web Request"
    },
    "operation": {
      "class": "OCA\\Talk\\Flow\\Operation",
      "name": "Write to conversation"
    },
    "description": "Flow activation: rules were requested"
  },
  "userAgent": "curl/7.68.0",
  "version": "18.0.0.8"
}
{
  "reqId": "X5XP4vROpcGCoaE6Ria2",
  "level": 1,
  "time": "2020-01-15T09:32:29+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "master",
  "app": "workflowengine",
  "method": "GET",
  "url": "/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping",
  "message": {
    "entity": {
      "class": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
      "name": "Web Request"
    },
    "operation": {
      "class": "OCA\\Talk\\Flow\\Operation",
      "name": "Write to conversation"
    },
    "configuration": {
      "id": 60,
      "class": "OCA\\Talk\\Flow\\Operation",
      "name": "",
      "checks": "[81]",
      "operation": "{\"m\":3,\"t\":\"wfwqkvsc\"}",
      "events": "[\"FlowHttpRequest::incomingRequest\"]",
      "entity": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
      "scope_type": 1,
      "scope_actor_id": "master"
    },
    "description": "Flow rule qualified to run"
  },
  "userAgent": "curl/7.68.0",
  "version": "18.0.0.8"
}
{
  "reqId": "X5XP4vROpcGCoaE6Ria2",
  "level": 1,
  "time": "2020-01-15T09:32:29+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "master",
  "app": "workflowengine",
  "method": "GET",
  "url": "/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping",
  "message": {
    "entity": {
      "class": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
      "name": "Web Request"
    },
    "operation": {
      "class": "OCA\\Talk\\Flow\\Operation",
      "name": "Write to conversation"
    },
    "configuration": [
      {
        "id": 60,
        "class": "OCA\\Talk\\Flow\\Operation",
        "name": "",
        "checks": "[81]",
        "operation": "{\"m\":3,\"t\":\"wfwqkvsc\"}",
        "events": "[\"FlowHttpRequest::incomingRequest\"]",
        "entity": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
        "scope_type": 1,
        "scope_actor_id": "master"
      }
    ],
    "description": "All qualified flow configurations are going to run"
  },
  "userAgent": "curl/7.68.0",
  "version": "18.0.0.8"
}
{
  "reqId": "X5XP4vROpcGCoaE6Ria2",
  "level": 1,
  "time": "2020-01-15T09:32:29+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "master",
  "app": "workflowengine",
  "method": "GET",
  "url": "/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping",
  "message": {
    "operation": {
      "class": "OCA\\Talk\\Flow\\Operation",
      "name": "Write to conversation"
    },
    "entity": {
      "class": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
      "name": "Web Request"
    },
    "eventName": "FlowHttpRequest::incomingRequest",
    "description": "Flow handling for event done"
  },
  "userAgent": "curl/7.68.0",
  "version": "18.0.0.8"
}
{
  "reqId": "X5XP4vROpcGCoaE6Ria2",
  "level": 1,
  "time": "2020-01-15T09:32:29+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "master",
  "app": "workflowengine",
  "method": "GET",
  "url": "/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping",
  "message": {
    "operation": {
      "class": "OCA\\WorkflowScript\\Operation",
      "name": "External scripts"
    },
    "entity": {
      "class": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
      "name": "Web Request"
    },
    "eventName": "FlowHttpRequest::incomingRequest",
    "description": "Flow activated by event"
  },
  "userAgent": "curl/7.68.0",
  "version": "18.0.0.8"
}
{
  "reqId": "X5XP4vROpcGCoaE6Ria2",
  "level": 1,
  "time": "2020-01-15T09:32:29+00:00",
  "remoteAddr": "127.0.0.1",
  "user": "master",
  "app": "workflowengine",
  "method": "GET",
  "url": "/master/ocs/v2.php/apps/flow_http_requests/api/v1/f/m7b43vgb?action=ping",
  "message": {
    "operation": {
      "class": "OCA\\WorkflowScript\\Operation",
      "name": "External scripts"
    },
    "entity": {
      "class": "OCA\\FlowHttpRequests\\Flow\\RequestEntity",
      "name": "Web Request"
    },
    "eventName": "FlowHttpRequest::incomingRequest",
    "description": "Flow handling for event done"
  },
  "userAgent": "curl/7.68.0",
  "version": "18.0.0.8"
}
```
</details>

### CODE REVIEW

File: apps/workflowengine/lib/Helper/LogContext.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+					default:
+						continue;
+				}
+			}
+		}
+		return $this;
+	}
+
+	public function setOperation(?IOperation $operation): LogContext {
+		if($operation instanceof IOperation) {
```
Comment (ChristophWurst) [inline apps/workflowengine/lib/Helper/LogContext.php]: given the type hint on the parameter this change is a bit weird. is this a null check?

also, why is the parameter nullable when passing in `null` doesn't actually do anything? :thinking:

File: apps/workflowengine/lib/Helper/LogContext.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+			$this->details['operation'] = [
+				'class' => get_class($operation),
+				'name' => $operation->getDisplayName(),
+			];
+		}
+		return $this;
+	}
+
+	public function setEntity(?IEntity $entity): LogContext {
+		if($entity instanceof IEntity) {
```
Comment (ChristophWurst) [inline apps/workflowengine/lib/Helper/LogContext.php]: same here

File: lib/public/Log/IDataLogger.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * @since 18.0.0
+ */
+interface IDataLogger {
+
+	/**
+	 * allows to log custom data, similar to how logException works
+	 *
+	 * @since 18.0.0
+	 */
+	public function logData(array $data, array $context = []): void;
```
Comment (ChristophWurst) [inline lib/public/Log/IDataLogger.php]: should we also add something like a `message` parameter here? all other logger methods either take a string or an exception that contains a message.

File: apps/workflowengine/lib/Helper/LogContext.php (hunk 1 above)
Comment (blizzz) [reply apps/workflowengine/lib/Helper/LogContext.php]: doesn't bloat the code in RuleMatcher to much, where it is not guaranteed that an operation is set (you can work around with using deprecated API)

Comment (blizzz) [reply apps/workflowengine/lib/Helper/LogContext.php]: If nothing is set, log should not crash

File: apps/workflowengine/lib/Helper/LogContext.php (hunk 2 above)
Comment (blizzz) [reply apps/workflowengine/lib/Helper/LogContext.php]: see above

File: lib/public/Log/IDataLogger.php (hunk 3 above)
Comment (blizzz) [reply lib/public/Log/IDataLogger.php]: imho this should be just part of the data array that is being passed into

**Comment by blizzz** [comment]
@ChristophWurst ok?

others – reviews :sunflower:  ?

**Comment by icewind1991** [comment]
If this has a chance of being logged to the main log the the logreader will need to be able to handle it

**Comment by blizzz** [comment]
@icewind1991 are Exceptions handled specifically or in a generic way?

**Comment by icewind1991** [comment]
They are handled specifically to have them fancy formatted.

For messages like this, having a guaranteed `message` field being set that can be shown should be enough, with the rest of the data just being json formatted

**Review by juliusknorr — APPROVED** [review]

**Comment by blizzz** [comment]
> If this has a chance of being logged to the main log the the logreader will need to be able to handle it

- in general, it is if some other apps would to this
- there is, if you set the logfile location to be the nextcloud log file
  - the logreader is crashing in that case

**Comment by icewind1991** [comment]
I would prefer to have a `string $message` parameter though so we always have something human readable in a predictable location

**Comment by blizzz** [comment]
A message parameter is not enough for the logreader, since it just tries to print the delivered payload if it does not match any Exception pattern.

What works better is to split the payload into a message and data part. Then the log reader works as it is now (extra fields are not being show but can be copied rawly).

I don't like the transformation kung foo n LogDetails however, but that's necessary to another field to the log output. Back to review.

File: apps/workflowengine/lib/Helper/LogContext.php (hunk 1 above)
Comment (ChristophWurst) [reply apps/workflowengine/lib/Helper/LogContext.php]: let me rephrase: did you mean to `if ($operation !== null)`? Because except for `null` the `$operation` will always be of type `IOperation` as per type hint.

Comment (blizzz) [reply apps/workflowengine/lib/Helper/LogContext.php]: yes

**Comment by blizzz** [comment]
are we fine with this now? @icewind1991 @rullzer  @ChristophWurst

File: lib/public/Log/IDataLogger.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ *
+ */
+
+namespace OCP\Log;
+
+/**
+ * Interface IDataLogger
+ *
+ * @package OCP\Log
+ * @since 18.0.1
```
Comment (ChristophWurst) [inline lib/public/Log/IDataLogger.php]: 19 :wink:

**Review by ChristophWurst — APPROVED** [review]

File: lib/public/Log/IDataLogger.php (hunk 4 above)
Comment (rullzer) [reply lib/public/Log/IDataLogger.php]: Will be back ported
